nodeJS/express/controllers.md

```diff
@@ -0,0 +1,466 @@
+1. With the string `next('router')` - Will skip all middleware functions attached to the specific router instance and pass control back out of the router instance. Basically, we exit the router and go back to the parent router, e.g. `app` (yes, the Express app is also just a router under the hood).
+
+Out of the four, you will likely only use the first two, unless you have a very specific need that requires the other two.
+
+### Consolidating what we've learned
+
+The following is the folder structure you'd end up with if you have been coding along:
+
+```text
+express-app/
+├─ controllers/
+│  ├─ userController.js
+├─ routes/
+│  ├─ userRoutes.js
+├─ app.js
+```
+
+Creating a user controller file that contains well-defined responsibilities to handle "User".
+
+```javascript
+// user controller file - controllers/userController.js
+
+const asyncHandler = require("express-async-handler");
+
+const getUserById = asyncHandler(async (req, res) => {
+  const userId = req.params.id;
+
+  const user = await someDBQueryToGetUser(userId);
+
+  if (!user) {
+    throw new CustomNotFoundError("User not found");
+  }
+
+  res.send(`User found: ${user.name}`);
+});
+
+const getUsers = asyncHandler(async (req, res) => {
+  // code
+});
+
+const createUser = asyncHandler(async (req, res) => {
+  // code
+});
+
+module.exports = { getUserById, getUsers, createUser };
+```
+
+Route file for the "User"
+
+```javascript
+// route file - routes/userRoutes.js
+
+const express = require('express');
+const userController = require('<path-to-user-controller>');
+
+const router = express.Router();
+
+// You can for example add a top level middleware function that handles say authentication and only let the request come in if they're authenticated
+// This prevents from executing the middleware functions below if the request is not authenticated
+// We will learn more about authentication in later lessons
+// usually calls either next() or next(error)
+
+// router.use(authMiddleware);
+
+// router-level midlewares
+
+// GET request for getting all the users
+router.get('/', userController.getUsers);
+
+// You will likely place your validation/authentication middleware functions here or perhaps in the controller file, e.g.
+// router.post(validationMiddleware, userController.createUser)
+
+// POST request for creating a user
+router.post('/', userController.createUser);
+
+// GET request for getting the user by id
+router.get('/:id', userControoler.getUserById);
+
+module.exports = router;
+```
+
+The main application file, our entry point.
+
+```javascript
+// app file
+
+const express = require('express');
+const morgan = require('morgan');
+const userRouter = require('<path-to-user-routes>');
+
+const app = express();
+
+// application-level middlewares, will always execute on every incoming requests, base mount path is `/`
+
+// a nice to have logger, will automatically show logs for every http request
+app.use(morgan('dev'));
```

Comment:
Can we remove logger from this code example? This lesson is already super long. I'll add a logger recommendation in one of the project lessons (either mini message board or inventory app)

Reply:
done